core/types/calls.cc

```diff
@@ -3320,6 +3320,25 @@ class Tuple_concat : public IntrinsicMethod {
     }
 } Tuple_concat;
 
+class Tuple_plus : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        auto *tuple = cast_type<TupleType>(args.thisType);
+        ENFORCE(tuple != nullptr);
+
+        auto specialEmptyTupleCase = tuple->elems.empty() && args.args.size() == 1 &&
+                                     (isa_type<TupleType>(args.args[0]->type) ||
+                                      Types::isSubType(gs, Types::arrayOfUntyped(), args.args[0]->type));
```

Comment:
this should guarantee that arg0 is an array (and not even an array subclass, so we don't have to worry about doing something like getting the `appliedType->klass` from the receiver and doing an `alignBaseTypeArgs` or something.